x-pack/metricbeat/module/gcp/metrics/redis/metadata.go

```diff
@@ -0,0 +1,210 @@
+	if err != nil {
+		return gcp.MetadataCollectorData{}, err
+	}
+
+	if resp.Resource != nil && resp.Resource.Labels != nil {
+		_, _ = metadataCollectorData.ECS.Put(gcp.ECSCloudInstanceIDKey, resp.Resource.Labels[gcp.TimeSeriesResponsePathForECSInstanceID])
+	}
+
+	_, _ = metadataCollectorData.ECS.Put(gcp.ECSCloudInstanceNameKey, metadata.instanceName)
+
+	if metadata.machineType != "" {
+		lastIndex := strings.LastIndex(metadata.machineType, "/")
+		_, _ = metadataCollectorData.ECS.Put(gcp.ECSCloudMachineTypeKey, metadata.machineType[lastIndex+1:])
+	}
+
+	metadata.Metrics = metadataCollectorData.Labels[gcp.LabelMetrics]
+	metadata.System = metadataCollectorData.Labels[gcp.LabelSystem]
+
+	if metadata.User != nil {
+		metadataCollectorData.Labels[gcp.LabelUser] = metadata.User
+	}
+
+	return metadataCollectorData, nil
+}
+
+// instanceMetadata returns the labels of an instance
+func (s *metadataCollector) instanceMetadata(ctx context.Context, instanceID, region string) (*redisMetadata, error) {
+	instance, err := s.instance(ctx, instanceID)
+	if err != nil {
+		return nil, fmt.Errorf("error trying to get data from instance '%s' %w", instanceID, err)
+	}
+
+	metadata := &redisMetadata{
+		instanceID: instanceID,
+		region:     region,
+	}
+
+	if instance == nil {
+		s.logger.Debugf("couldn't get instance '%s' call ListInstances API", instanceID)
+		return metadata, nil
+	}
+
+	if instance.Name != "" {
+		parts := strings.Split(instance.Name, "/")
+		metadata.instanceName = parts[len(parts)-1]
+	}
+
+	if instance.Labels != nil {
+		metadata.User = instance.Labels
+	}
+
+	if instance.Tier.String() != "" {
+		metadata.machineType = instance.Tier.String()
+	}
+
+	return metadata, nil
+}
+
+// instance returns data from an instance ID using the cache or making a request
+func (s *metadataCollector) instance(ctx context.Context, instanceID string) (*redispb.Instance, error) {
+	s.getInstances(ctx)
+
+	instance, ok := s.instances[instanceID]
+	if ok {
+		return instance, nil
+	}
+
+	s.instances = make(map[string]*redispb.Instance)
+
+	return nil, nil
+}
+
+func (s *metadataCollector) instanceID(ts *monitoringpb.TimeSeries) string {
+	if ts.Resource != nil && ts.Resource.Labels != nil {
+		return ts.Resource.Labels[gcp.TimeSeriesResponsePathForECSInstanceID]
+	}
+
+	return ""
+}
+
+func (s *metadataCollector) instanceRegion(ts *monitoringpb.TimeSeries) string {
+	if ts.Resource != nil && ts.Resource.Labels != nil {
+		return ts.Resource.Labels["region"]
+	}
+
+	return ""
+}
+
+func (s *metadataCollector) getInstances(ctx context.Context) {
+	if len(s.instances) > 0 {
+		return
+	}
+
+	s.logger.Debug("get redis instances with ListInstances API")
+
+	client, err := redisa.NewCloudRedisClient(ctx, s.opt...)
+	if err != nil {
+		s.logger.Errorf("error getting client from redis service: %v", err)
+		return
+	}
+
+	defer client.Close()
+
+	// Use locations - (wildcard) to fetch all instances.
```

Comment:
I've not been able to find documentation for this wildcard. Do you have a link that we could add here to document its usage?

Reply:
Found that from the official documentation of the SDK.

https://pkg.go.dev/cloud.google.com/go/redis@v1.10.0/apiv1#CloudRedisClient.ListInstances

`If location_id is specified as - (wildcard), then all regions available to the project are queried, and the results are aggregated. `

Reply:
Can you add the link in the code?